### ISSUE TITLE
Proper benchmarking with JMH

### ISSUE THREAD

**Issue by LunNova** [nallar/WhoCalled#1]
The current code hasn't been benchmarked properly.

I tested a previous version of it (only changes since then are rename refactoring and moving into a separate project) in TickThreading: [ContextAccess.java](https://github.com/nallar/TickThreading/blob/1.6.4/src/main/java/nallar/tickthreading/util/contextaccess/ContextAccess.java)

I confirmed that using `SecurityManager`'s getContext is faster than using `Thread.currentThread().getStackTrace()`, however it was tested by swapping implementations on a production server and using a sampling profiler. This isn't reliable.

**Comment by LunNova** [comment]
```
Benchmark                                               Mode  Samples     Score    Error  Units
m.n.w.WhoCalledBenchmark.testReflectionCalledBy         avgt       10  2178.268 ± 86.156  ns/op
m.n.w.WhoCalledBenchmark.testReflectionGet              avgt       10    86.975 ±  5.302  ns/op
m.n.w.WhoCalledBenchmark.testSecurityManagerCalledBy    avgt       10   495.695 ± 12.243  ns/op
m.n.w.WhoCalledBenchmark.testSecurityManagerGet         avgt       10   502.327 ± 22.478  ns/op
m.n.w.WhoCalledBenchmark.testStackTraceCalledBy         avgt       10  8630.241 ± 21.425  ns/op
m.n.w.WhoCalledBenchmark.testStackTraceGet              avgt       10  9161.564 ± 85.620  ns/op
```

**Comment by LunNova** [comment]
Due to these results I have changed WhoCalledReflection to use the SecurityManager.getClassContext method to implement isCalledBy, as that is faster and always available.

WhoCalledStackTrace is only there for testing purposes, and should never actually be selected as WhoCalledSecurityManager will always work.

**Comment by LunNova** [comment]
A small performance increase for normal usage has been added as described in [WhoCalledProvider.java](https://github.com/nallar/WhoCalled/blob/master/src/main/java/me/nallar/whocalled/WhoCalledProvider.java#L11):

> For performance reasons, we list the first java version an implementation will not work on.
> By avoiding loading more than one implementation of the WhoCalled interface, the jit can produce slightly faster devirtualized calls. We never get a virtual call when both classes are loaded, but a very slight overhead is added.
> Instead of non-virtual call, we get non-virtual call with check that class matches.
> 
> This is discussed in detail in the following talk:
> JVM Mechanics - Silicon Valley JUG 2015 at https://www.youtube.com/watch?v=E9i9NJeXGmM

Implemented in commit 4147ed782204
